src/graph.rs

```diff
@@ -14,48 +14,106 @@ pub struct Graph<T> {
 
 impl<T: Eq + PartialEq + Hash> Graph<T> {
     pub fn new() -> Graph<T> {
-        todo!();
+        Graph { edges: HashMap::<Rc<T>, HashSet<Rc<T>>>::new(), }
     }
 
     pub fn vertices(&self) -> Vec<Rc<T>> {
-        todo!();
+        let mut vertices: HashSet<Rc<T>> = HashSet::new();
+        for (k, v) in self.edges.iter() {
+            vertices.insert(k.clone());
+            vertices.extend(v.iter().map(|x| x.clone()));
+        }
+        vertices.into_iter().collect()
     }
 
     pub fn insert_vertex(&mut self, u: T) {
-        todo!();
+        // if key is not present, insert key
+        if !self.edges.contains_key(&u) {
+            self.edges.insert(u.into(), HashSet::<Rc<T>>::new());
+        }
     }
 
     pub fn insert_edge(&mut self, u: T, v: T) {
-        // node u can already be in the HashMap or it is not in the HashMap
-        todo!();
+        // node is not present
+        if !self.edges.contains_key(&u) {
+            self.edges.insert(u.into(), HashSet::<Rc<T>>::from([v.into()]));
+        } else { // node is present
+            if let Some(hash_set) = self.edges.get_mut(&u) {
+                hash_set.insert(v.into());
+            }
+        }
     }
 
     pub fn remove_edge(&mut self, u: &T, v: &T) {
-        todo!();
+        self.edges.get_mut(u).map(|h| h.remove(v));
     }
 
     pub fn remove_vertex(&mut self, u: &T) {
```

Comment:
What about the vertex which has zero outgoing edge?
Given a graph A->B. Remove B.